src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,492 @@
+                    {
+                        return;
+                    }
+
+                    // All operation blocks for a symbol belong to the same tree.
+                    var firstBlock = context.OperationBlocks[0];
+                    if (!symbolStartAnalyzer._compilationAnalyzer.TryGetOptions(firstBlock.Syntax.SyntaxTree, firstBlock.Language,
+                        context.Options, context.CancellationToken, out var options))
+                    {
+                        return;
+                    }
+
+                    var blockAnalyzer = new BlockAnalyzer(symbolStartAnalyzer, options);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeExpressionStatement, OperationKind.ExpressionStatement);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeDelegateCreationOrAnonymousFunction, OperationKind.DelegateCreation, OperationKind.AnonymousFunction);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeConversion, OperationKind.Conversion);
+                    context.RegisterOperationAction(blockAnalyzer.AnalyzeFieldOrPropertyReference, OperationKind.FieldReference, OperationKind.PropertyReference);
+                    context.RegisterOperationBlockEndAction(blockAnalyzer.AnalyzeOperationBlockEnd);
+
+                    return;
+
+                    // Local Functions.
+                    bool HasSyntaxErrors()
+                    {
+                        foreach (var operationBlock in context.OperationBlocks)
+                        {
+                            if (operationBlock.SemanticModel.GetSyntaxDiagnostics(operationBlock.Syntax.Span, context.CancellationToken).HasAnyErrors())
+                            {
+                                return true;
+                            }
+                        }
+
+                        return false;
+                    }
+                }
+
+                private void AnalyzeExpressionStatement(OperationAnalysisContext context)
+                {
+                    if (_options.UnusedValueExpressionStatementSeverity == ReportDiagnostic.Suppress)
+                    {
+                        return;
+                    }
+
+                    var expressionStatement = (IExpressionStatementOperation)context.Operation;
+                    var value = expressionStatement.Operation;
+
+                    // Bail out cases for report unused expression value:
+
+                    //  1. Null type and void returning method invocations: no value being dropped here.
+                    if (value.Type == null ||
+                        value.Type.SpecialType == SpecialType.System_Void)
+                    {
+                        return;
+                    }
+
+                    //  2. Bail out for syntax error (constant expressions) and semantic error (invalid operation) cases.
+                    if (value.ConstantValue.HasValue ||
```

Comment:
i don't understand what htis means: " Bail out for syntax error (constant expressions) "

how is a constant expression a syntax error?

Reply:
Changed the comment to:

```
//  2. Bail out for semantic error (invalid operation) cases.
//     Also bail out for constant expressions in expression statement syntax, say as "1;",
//     which do not seem to have an invalid operation in the operation tree.
```

I also have a unit test `ExpressionStatement_ConstantValue` for this scenario, which fails if we don't bail out on constant values explicitly.

---
In reply to: [234802357](https://github.com/dotnet/roslyn/pull/30777#discussion_r234802357) [](ancestors = 234802357)

Reply:
Great.  Thanks for hte clarification!